service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +461,233 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+
+                  if (assertRefSnapshotId == null || setSnapshotRef == null) {
+                    // This implies the request was not trying to add a snapshot.
+                    LOGGER.debug(
+                        "Giving up on Rollback replace operations for table={}, with current-snapshot-id={}, as operation doesn't attempts to add a single snapshot",
+                        base.uuid(),
+                        base.currentSnapshot().snapshotId());
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // snapshot-id the client expects the table current_snapshot_id
+                  long expectedCurrentSnapshotId = assertRefSnapshotId.snapshotId();
+
+                  MetadataUpdate.AddSnapshot snapshotToBeAdded = findAddSnapshotUpdate(request);
+                  if (snapshotToBeAdded == null) {
+                    // Re-throw if, there's no snapshot data to be added.
+                    throw new ValidationFailureException(e);
+                  }
+
+                  LOGGER.info(
+                      "Attempting to Rollback replace operation for table={}, with current-snapshot-id={}, to snapshot={}",
+                      base.uuid(),
+                      base.currentSnapshot().snapshotId(),
+                      snapshotToBeAdded.snapshot().snapshotId());
+
+                  List<MetadataUpdate> metadataUpdates =
+                      generateUpdatesToRemoveNoopSnapshot(
+                          base, expectedCurrentSnapshotId, setSnapshotRef.name());
+
+                  if (metadataUpdates == null || metadataUpdates.isEmpty()) {
+                    // Nothing can be done as this implies that there were not all
+                    // No-op snapshots (REPLACE) between expectedCurrentSnapshotId and
+                    // currentSnapshotId. hence re-throw the exception caught.
+                    throw new ValidationFailureException(e);
+                  }
+                  // Set back the ref we wanted to set, back to the snapshot-id
+                  // the client is expecting the table to be at.
+                  metadataBuilder.setBranchSnapshot(
+                      expectedCurrentSnapshotId, setSnapshotRef.name());
+
+                  // apply the remove snapshots update in the current metadata.
+                  // NOTE: we need to setRef to expectedCurrentSnapshotId first and then apply
+                  // remove, as otherwise the remove will drop the reference.
+                  // NOTE: we can skip removing the now orphan base. Its not a hard requirement.
+                  // just something good to do, and not leave for Remove Orphans.
+                  // Ref rolled back update correctly to snapshot to be committed parent now.
+                  metadataUpdates.forEach((update -> update.applyTo(metadataBuilder)));
+                  newBase = setAppropriateLastSeqNumber(metadataBuilder.build());
```

Comment:
could we set `lastSequenceNumber` in the _builder_ instead? I think that would be preferable from the java language perspective because that field is not `final`. Also, exposing a setter in the builder upstream is more natural.